### ISSUE TITLE
Test failures on qiskit-terra>=0.14.0: test_random_unitary_gate

### ISSUE THREAD

**Issue by drewrisinger** [Qiskit/qiskit-aer#731]
### Informations

- **Qiskit Aer version**: 0.5.1
- **Qiskit Terra version**: 0.14.[0,1]
- **Python version**: 3.7
- **Operating system**: Linux (Nix)

### What is the current behavior?

When running tests via pytest:
<details>
<summary>Test log from failed qiskit-aer tests</summary>

```
__________________ TestQasmSimulator.test_random_unitary_gate __________________

self = <test.terra.backends.test_qasm_simulator.TestQasmSimulator testMethod=test_random_unitary_gate>

    def test_random_unitary_gate(self):
        """Test simulation with random unitary gate circuit instructions."""
        shots = 2000
        circuits = ref_unitary_gate.unitary_random_gate_circuits_nondeterministic(final_measure=True)
        targets = ref_unitary_gate.unitary_random_gate_counts_nondeterministic()
        result = execute(circuits, self.SIMULATOR, shots=shots).result()
        self.assertTrue(getattr(result, 'success', False))
>       self.compare_counts(result, circuits, targets, delta=0.05 * shots)

test/terra/backends/qasm_simulator/qasm_unitary_gate.py:52: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/terra/common.py:169: in compare_counts
    output, target, delta=delta, msg=msg)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <test.terra.backends.test_qasm_simulator.TestQasmSimulator testMethod=test_random_unitary_gate>
dict1 = {'0x0': 21, '0x1': 1658, '0x2': 185, '0x3': 136}
dict2 = {'0x0': 930, '0x1': 330, '0x2': 680, '0x3': 60}, delta = 100.0
msg = "('0x1': 1658 != 330), ('0x2': 185 != 680), ('0x0': 21 != 930) within 100.0 delta :  {'0x3': 136, '0x1': 1658, '0x2': 185, '0x0': 21} != {'0x0': 930, '0x1': 330, '0x2': 680, '0x3': 60}"
places = None, default_value = 0

    def assertDictAlmostEqual(self, dict1, dict2, delta=None, msg=None,
                              places=None, default_value=0):
        """
        Assert two dictionaries with numeric values are almost equal.
    
        Fail if the two dictionaries are unequal as determined by
        comparing that the difference between values with the same key are
        not greater than delta (default 1e-8), or that difference rounded
        to the given number of decimal places is not zero. If a key in one
        dictionary is not in the other the default_value keyword argument
        will be used for the missing value (default 0). If the two objects
        compare equal then they will automatically compare almost equal.
    
        Args:
            dict1 (dict): a dictionary.
            dict2 (dict): a dictionary.
            delta (number): threshold for comparison (defaults to 1e-8).
            msg (str): return a custom message on failure.
            places (int): number of decimal places for comparison.
            default_value (number): default value for missing keys.
    
        Raises:
            TypeError: raises TestCase failureException if the test fails.
        """
        if dict1 == dict2:
            # Shortcut
            return
        if delta is not None and places is not None:
            raise TypeError("specify delta or places not both")
    
        if places is not None:
            success = True
            standard_msg = ''
            # check value for keys in target
            keys1 = set(dict1.keys())
            for key in keys1:
                val1 = dict1.get(key, default_value)
                val2 = dict2.get(key, default_value)
                if round(abs(val1 - val2), places) != 0:
                    success = False
                    standard_msg += '(%s: %s != %s), ' % (safe_repr(key),
                                                          safe_repr(val1),
                                                          safe_repr(val2))
            # check values for keys in counts, not in target
            keys2 = set(dict2.keys()) - keys1
            for key in keys2:
                val1 = dict1.get(key, default_value)
                val2 = dict2.get(key, default_value)
                if round(abs(val1 - val2), places) != 0:
                    success = False
                    standard_msg += '(%s: %s != %s), ' % (safe_repr(key),
                                                          safe_repr(val1),
                                                          safe_repr(val2))
            if success is True:
                return
            standard_msg = standard_msg[:-2] + ' within %s places' % places
    
        else:
            if delta is None:
                delta = 1e-8  # default delta value
            success = True
            standard_msg = ''
            # check value for keys in target
            keys1 = set(dict1.keys())
            for key in keys1:
                val1 = dict1.get(key, default_value)
                val2 = dict2.get(key, default_value)
                if abs(val1 - val2) > delta:
                    success = False
                    standard_msg += '(%s: %s != %s), ' % (safe_repr(key),
                                                          safe_repr(val1),
                                                          safe_repr(val2))
            # check values for keys in counts, not in target
            keys2 = set(dict2.keys()) - keys1
            for key in keys2:
                val1 = dict1.get(key, default_value)
                val2 = dict2.get(key, default_value)
                if abs(val1 - val2) > delta:
                    success = False
                    standard_msg += '(%s: %s != %s), ' % (safe_repr(key),
                                                          safe_repr(val1),
                                                          safe_repr(val2))
            if success is True:
                return
            standard_msg = standard_msg[:-2] + ' within %s delta' % delta
    
        msg = self._formatMessage(msg, standard_msg)
>       raise self.failureException(msg)
E       AssertionError: ('0x1': 1658 != 330), ('0x2': 185 != 680), ('0x0': 21 != 930) within 100.0 delta :  {'0x3': 136, '0x1': 1658, '0x2': 185, '0x0': 21} != {'0x0': 930, '0x1': 330, '0x2': 680, '0x3': 60}

test/terra/common.py:289: AssertionError
___________ TestQasmSimulatorDensityMatrix.test_random_unitary_gate ____________

self = <test.terra.backends.test_qasm_simulator_density_matrix.TestQasmSimulatorDensityMatrix testMethod=test_random_unitary_gate>

    def test_random_unitary_gate(self):
        """Test simulation with random unitary gate circuit instructions."""
        shots = 2000
        circuits = ref_unitary_gate.unitary_random_gate_circuits_nondeterministic(final_measure=True)
        targets = ref_unitary_gate.unitary_random_gate_counts_nondeterministic()
        result = execute(circuits, self.SIMULATOR, shots=shots).result()
        self.assertTrue(getattr(result, 'success', False))
>       self.compare_counts(result, circuits, targets, delta=0.05 * shots)

test/terra/backends/qasm_simulator/qasm_unitary_gate.py:52: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/terra/common.py:169: in compare_counts
    output, target, delta=delta, msg=msg)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <test.terra.backends.test_qasm_simulator_density_matrix.TestQasmSimulatorDensityMatrix testMethod=test_random_unitary_gate>
dict1 = {'0x0': 25, '0x1': 1634, '0x2': 215, '0x3': 126}
dict2 = {'0x0': 930, '0x1': 330, '0x2': 680, '0x3': 60}, delta = 100.0
msg = "('0x1': 1634 != 330), ('0x2': 215 != 680), ('0x0': 25 != 930) within 100.0 delta :  {'0x3': 126, '0x1': 1634, '0x2': 215, '0x0': 25} != {'0x0': 930, '0x1': 330, '0x2': 680, '0x3': 60}"
places = None, default_value = 0

    def assertDictAlmostEqual(self, dict1, dict2, delta=None, msg=None,
                              places=None, default_value=0):
        """
        Assert two dictionaries with numeric values are almost equal.
    
        Fail if the two dictionaries are unequal as determined by
        comparing that the difference between values with the same key are
        not greater than delta (default 1e-8), or that difference rounded
        to the given number of decimal places is not zero. If a key in one
        dictionary is not in the other the default_value keyword argument
        will be used for the missing value (default 0). If the two objects
        compare equal then they will automatically compare almost equal.
    
        Args:
            dict1 (dict): a dictionary.
            dict2 (dict): a dictionary.
            delta (number): threshold for comparison (defaults to 1e-8).
            msg (str): return a custom message on failure.
            places (int): number of decimal places for comparison.
            default_value (number): default value for missing keys.
    
        Raises:
            TypeError: raises TestCase failureException if the test fails.
        """
        if dict1 == dict2:
            # Shortcut
            return
        if delta is not None and places is not None:
            raise TypeError("specify delta or places not both")
    
        if places is not None:
            success = True
            standard_msg = ''
            # check value for keys in target
            keys1 = set(dict1.keys())
            for key in keys1:
                val1 = dict1.get(key, default_value)
                val2 = dict2.get(key, default_value)
                if round(abs(val1 - val2), places) != 0:
                    success = False
                    standard_msg += '(%s: %s != %s), ' % (safe_repr(key),
                                                          safe_repr(val1),
                                                          safe_repr(val2))
            # check values for keys in counts, not in target
            keys2 = set(dict2.keys()) - keys1
            for key in keys2:
                val1 = dict1.get(key, default_value)
                val2 = dict2.get(key, default_value)
                if round(abs(val1 - val2), places) != 0:
                    success = False
                    standard_msg += '(%s: %s != %s), ' % (safe_repr(key),
                                                          safe_repr(val1),
                                                          safe_repr(val2))
            if success is True:
                return
            standard_msg = standard_msg[:-2] + ' within %s places' % places
    
        else:
            if delta is None:
                delta = 1e-8  # default delta value
            success = True
            standard_msg = ''
            # check value for keys in target
            keys1 = set(dict1.keys())
            for key in keys1:
                val1 = dict1.get(key, default_value)
                val2 = dict2.get(key, default_value)
                if abs(val1 - val2) > delta:
                    success = False
                    standard_msg += '(%s: %s != %s), ' % (safe_repr(key),
                                                          safe_repr(val1),
                                                          safe_repr(val2))
            # check values for keys in counts, not in target
            keys2 = set(dict2.keys()) - keys1
            for key in keys2:
                val1 = dict1.get(key, default_value)
                val2 = dict2.get(key, default_value)
                if abs(val1 - val2) > delta:
                    success = False
                    standard_msg += '(%s: %s != %s), ' % (safe_repr(key),
                                                          safe_repr(val1),
                                                          safe_repr(val2))
            if success is True:
                return
            standard_msg = standard_msg[:-2] + ' within %s delta' % delta
    
        msg = self._formatMessage(msg, standard_msg)
>       raise self.failureException(msg)
E       AssertionError: ('0x1': 1634 != 330), ('0x2': 215 != 680), ('0x0': 25 != 930) within 100.0 delta :  {'0x3': 126, '0x1': 1634, '0x2': 215, '0x0': 25} != {'0x0': 930, '0x1': 330, '0x2': 680, '0x3': 60}

test/terra/common.py:289: AssertionError
_________ TestQasmMatrixProductStateSimulator.test_random_unitary_gate _________

self = <test.terra.backends.test_qasm_simulator_matrix_product_state.TestQasmMatrixProductStateSimulator testMethod=test_random_unitary_gate>

    def test_random_unitary_gate(self):
        """Test simulation with random unitary gate circuit instructions."""
        shots = 2000
        circuits = ref_unitary_gate.unitary_random_gate_circuits_nondeterministic(final_measure=True)
        targets = ref_unitary_gate.unitary_random_gate_counts_nondeterministic()
        result = execute(circuits, self.SIMULATOR, shots=shots).result()
        self.assertTrue(getattr(result, 'success', False))
>       self.compare_counts(result, circuits, targets, delta=0.05 * shots)

test/terra/backends/qasm_simulator/qasm_unitary_gate.py:52: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/terra/common.py:169: in compare_counts
    output, target, delta=delta, msg=msg)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <test.terra.backends.test_qasm_simulator_matrix_product_state.TestQasmMatrixProductStateSimulator testMethod=test_random_unitary_gate>
dict1 = {'0x0': 22, '0x1': 1655, '0x2': 176, '0x3': 147}
dict2 = {'0x0': 930, '0x1': 330, '0x2': 680, '0x3': 60}, delta = 100.0
msg = "('0x1': 1655 != 330), ('0x2': 176 != 680), ('0x0': 22 != 930) within 100.0 delta :  {'0x3': 147, '0x1': 1655, '0x2': 176, '0x0': 22} != {'0x0': 930, '0x1': 330, '0x2': 680, '0x3': 60}"
places = None, default_value = 0

    def assertDictAlmostEqual(self, dict1, dict2, delta=None, msg=None,
                              places=None, default_value=0):
        """
        Assert two dictionaries with numeric values are almost equal.
    
        Fail if the two dictionaries are unequal as determined by
        comparing that the difference between values with the same key are
        not greater than delta (default 1e-8), or that difference rounded
        to the given number of decimal places is not zero. If a key in one
        dictionary is not in the other the default_value keyword argument
        will be used for the missing value (default 0). If the two objects
        compare equal then they will automatically compare almost equal.
    
        Args:
            dict1 (dict): a dictionary.
            dict2 (dict): a dictionary.
            delta (number): threshold for comparison (defaults to 1e-8).
            msg (str): return a custom message on failure.
            places (int): number of decimal places for comparison.
            default_value (number): default value for missing keys.
    
        Raises:
            TypeError: raises TestCase failureException if the test fails.
        """
        if dict1 == dict2:
            # Shortcut
            return
        if delta is not None and places is not None:
            raise TypeError("specify delta or places not both")
    
        if places is not None:
            success = True
            standard_msg = ''
            # check value for keys in target
            keys1 = set(dict1.keys())
            for key in keys1:
                val1 = dict1.get(key, default_value)
                val2 = dict2.get(key, default_value)
                if round(abs(val1 - val2), places) != 0:
                    success = False
                    standard_msg += '(%s: %s != %s), ' % (safe_repr(key),
                                                          safe_repr(val1),
                                                          safe_repr(val2))
            # check values for keys in counts, not in target
            keys2 = set(dict2.keys()) - keys1
            for key in keys2:
                val1 = dict1.get(key, default_value)
                val2 = dict2.get(key, default_value)
                if round(abs(val1 - val2), places) != 0:
                    success = False
                    standard_msg += '(%s: %s != %s), ' % (safe_repr(key),
                                                          safe_repr(val1),
                                                          safe_repr(val2))
            if success is True:
                return
            standard_msg = standard_msg[:-2] + ' within %s places' % places
    
        else:
            if delta is None:
                delta = 1e-8  # default delta value
            success = True
            standard_msg = ''
            # check value for keys in target
            keys1 = set(dict1.keys())
            for key in keys1:
                val1 = dict1.get(key, default_value)
                val2 = dict2.get(key, default_value)
                if abs(val1 - val2) > delta:
                    success = False
                    standard_msg += '(%s: %s != %s), ' % (safe_repr(key),
                                                          safe_repr(val1),
                                                          safe_repr(val2))
            # check values for keys in counts, not in target
            keys2 = set(dict2.keys()) - keys1
            for key in keys2:
                val1 = dict1.get(key, default_value)
                val2 = dict2.get(key, default_value)
                if abs(val1 - val2) > delta:
                    success = False
                    standard_msg += '(%s: %s != %s), ' % (safe_repr(key),
                                                          safe_repr(val1),
                                                          safe_repr(val2))
            if success is True:
                return
            standard_msg = standard_msg[:-2] + ' within %s delta' % delta
    
        msg = self._formatMessage(msg, standard_msg)
>       raise self.failureException(msg)
E       AssertionError: ('0x1': 1655 != 330), ('0x2': 176 != 680), ('0x0': 22 != 930) within 100.0 delta :  {'0x3': 147, '0x1': 1655, '0x2': 176, '0x0': 22} != {'0x0': 930, '0x1': 330, '0x2': 680, '0x3': 60}

test/terra/common.py:289: AssertionError
____________ TestQasmSimulatorStatevector.test_random_unitary_gate _____________

self = <test.terra.backends.test_qasm_simulator_statevector.TestQasmSimulatorStatevector testMethod=test_random_unitary_gate>

    def test_random_unitary_gate(self):
        """Test simulation with random unitary gate circuit instructions."""
        shots = 2000
        circuits = ref_unitary_gate.unitary_random_gate_circuits_nondeterministic(final_measure=True)
        targets = ref_unitary_gate.unitary_random_gate_counts_nondeterministic()
        result = execute(circuits, self.SIMULATOR, shots=shots).result()
        self.assertTrue(getattr(result, 'success', False))
>       self.compare_counts(result, circuits, targets, delta=0.05 * shots)

test/terra/backends/qasm_simulator/qasm_unitary_gate.py:52: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/terra/common.py:169: in compare_counts
    output, target, delta=delta, msg=msg)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <test.terra.backends.test_qasm_simulator_statevector.TestQasmSimulatorStatevector testMethod=test_random_unitary_gate>
dict1 = {'0x0': 24, '0x1': 1655, '0x2': 196, '0x3': 125}
dict2 = {'0x0': 930, '0x1': 330, '0x2': 680, '0x3': 60}, delta = 100.0
msg = "('0x1': 1655 != 330), ('0x2': 196 != 680), ('0x0': 24 != 930) within 100.0 delta :  {'0x3': 125, '0x1': 1655, '0x2': 196, '0x0': 24} != {'0x0': 930, '0x1': 330, '0x2': 680, '0x3': 60}"
places = None, default_value = 0

    def assertDictAlmostEqual(self, dict1, dict2, delta=None, msg=None,
                              places=None, default_value=0):
        """
        Assert two dictionaries with numeric values are almost equal.
    
        Fail if the two dictionaries are unequal as determined by
        comparing that the difference between values with the same key are
        not greater than delta (default 1e-8), or that difference rounded
        to the given number of decimal places is not zero. If a key in one
        dictionary is not in the other the default_value keyword argument
        will be used for the missing value (default 0). If the two objects
        compare equal then they will automatically compare almost equal.
    
        Args:
            dict1 (dict): a dictionary.
            dict2 (dict): a dictionary.
            delta (number): threshold for comparison (defaults to 1e-8).
            msg (str): return a custom message on failure.
            places (int): number of decimal places for comparison.
            default_value (number): default value for missing keys.
    
        Raises:
            TypeError: raises TestCase failureException if the test fails.
        """
        if dict1 == dict2:
            # Shortcut
            return
        if delta is not None and places is not None:
            raise TypeError("specify delta or places not both")
    
        if places is not None:
            success = True
            standard_msg = ''
            # check value for keys in target
            keys1 = set(dict1.keys())
            for key in keys1:
                val1 = dict1.get(key, default_value)
                val2 = dict2.get(key, default_value)
                if round(abs(val1 - val2), places) != 0:
                    success = False
                    standard_msg += '(%s: %s != %s), ' % (safe_repr(key),
                                                          safe_repr(val1),
                                                          safe_repr(val2))
            # check values for keys in counts, not in target
            keys2 = set(dict2.keys()) - keys1
            for key in keys2:
                val1 = dict1.get(key, default_value)
                val2 = dict2.get(key, default_value)
                if round(abs(val1 - val2), places) != 0:
                    success = False
                    standard_msg += '(%s: %s != %s), ' % (safe_repr(key),
                                                          safe_repr(val1),
                                                          safe_repr(val2))
            if success is True:
                return
            standard_msg = standard_msg[:-2] + ' within %s places' % places
    
        else:
            if delta is None:
                delta = 1e-8  # default delta value
            success = True
            standard_msg = ''
            # check value for keys in target
            keys1 = set(dict1.keys())
            for key in keys1:
                val1 = dict1.get(key, default_value)
                val2 = dict2.get(key, default_value)
                if abs(val1 - val2) > delta:
                    success = False
                    standard_msg += '(%s: %s != %s), ' % (safe_repr(key),
                                                          safe_repr(val1),
                                                          safe_repr(val2))
            # check values for keys in counts, not in target
            keys2 = set(dict2.keys()) - keys1
            for key in keys2:
                val1 = dict1.get(key, default_value)
                val2 = dict2.get(key, default_value)
                if abs(val1 - val2) > delta:
                    success = False
                    standard_msg += '(%s: %s != %s), ' % (safe_repr(key),
                                                          safe_repr(val1),
                                                          safe_repr(val2))
            if success is True:
                return
            standard_msg = standard_msg[:-2] + ' within %s delta' % delta
    
        msg = self._formatMessage(msg, standard_msg)
>       raise self.failureException(msg)
E       AssertionError: ('0x1': 1655 != 330), ('0x2': 196 != 680), ('0x0': 24 != 930) within 100.0 delta :  {'0x3': 125, '0x1': 1655, '0x2': 196, '0x0': 24} != {'0x0': 930, '0x1': 330, '0x2': 680, '0x3': 60}

test/terra/common.py:289: AssertionError
===== 4 failed, 1040 passed, 940 skipped, 39 warnings in 188.28s (0:03:08) =====
```
</details>

### Steps to reproduce the problem

1. Create environment with ``qiskit-terra==0.14.0`` & ``qiskit-aer==0.5.1``.
2. Run ``pytest`` from ``qiskit-aer`` git checkout @ tag ``0.5.1``

Alternatively, using Nix:

1. Install Nix
2. Clone https://github.com/drewrisinger/nur-packages & checkout commit https://github.com/drewrisinger/nur-packages/commit/4eacb968085e98e79aba178f3b78352ae019cd28
3. ``nix-build -A python3Packages.qiskit -j4``

### What is the expected behavior?

All tests should pass.

**Comment by mtreinish** [comment]
This should be fixed by: https://github.com/Qiskit/qiskit-aer/pull/727

**Comment by drewrisinger** [comment]
Tested working when patched with #727. Thanks!
